flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveFunctionDefinitionFactory.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFEvaluator;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFResolver2;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDF;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDTF;
+import org.apache.hadoop.mapred.JobConf;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * Factory to create {@link FunctionDefinition} for Hive user defined functions.
+ */
+public class HiveFunctionDefinitionFactory implements FunctionDefinitionFactory {
+	private static final Logger LOG = LoggerFactory.getLogger(HiveFunctionDefinitionFactory.class);
+
+	private final String hiveVersion;
+
+	public HiveFunctionDefinitionFactory(HiveConf hiveConf) {
+		checkNotNull(hiveConf);
+
+		this.hiveVersion = new JobConf(hiveConf).get(HiveCatalogValidator.CATALOG_HIVE_VERSION, HiveShimLoader.getHiveVersion());
+	}
+
+	@Override
+	public FunctionDefinition createFunctionDefinition(String name, CatalogFunction catalogFunction) {
+
+		String functionClassName = catalogFunction.getClassName();
+
+		if (Boolean.valueOf(catalogFunction.getProperties().get(CatalogConfig.IS_GENERIC))) {
+			throw new TableException(
+				String.format("HiveFunctionDefinitionFactory does not support generic functions %s yet", name));
+		}
+
+		Class clazz;
+		try {
+			clazz = Thread.currentThread().getContextClassLoader().loadClass(functionClassName);
+
+			LOG.info("Successfully loaded Hive udf '{}' with class '{}'", name, functionClassName);
+		} catch (ClassNotFoundException e) {
+			throw new TableException(
+				String.format("Failed to initiate an instance of class %s.", functionClassName), e);
+		}
+
+		if (UDF.class.isAssignableFrom(clazz)) {
+			LOG.info("Transforming Hive function '{}' into a HiveSimpleUDF", name);
+
+			return new ScalarFunctionDefinition(
+				name,
+				new HiveSimpleUDF(new HiveFunctionWrapper<>(functionClassName))
+			);
+		} else if (GenericUDF.class.isAssignableFrom(clazz)) {
+			LOG.info("Transforming Hive function '{}' into a HiveGenericUDF", name);
+
+			return new ScalarFunctionDefinition(
+				name,
+				new HiveGenericUDF(new HiveFunctionWrapper<>(functionClassName))
+			);
+		} else if (GenericUDTF.class.isAssignableFrom(clazz)) {
+			LOG.info("Transforming Hive function '{}' into a HiveGenericUDTF", name);
+
+			HiveGenericUDTF udtf = new HiveGenericUDTF(new HiveFunctionWrapper<>(functionClassName));
+
+			return new TableFunctionDefinition(
+				name,
+				udtf,
+				GenericTypeInfo.of(Row.class)
+			);
+		} else if (GenericUDAFResolver2.class.isAssignableFrom(clazz)) {
+			LOG.info(
+				"Transforming Hive function '{}' into a HiveGenericUDAF with no UDAF bridging and Hive version %s",
+				name, hiveVersion);
+
+			HiveGenericUDAF udaf = new HiveGenericUDAF(new HiveFunctionWrapper<>(functionClassName), false, hiveVersion);
```

Comment:
Minor: Extract a method to construct `HiveGenericUDAF` and `AggregateFunctionDefinition`?